lib/TimeSeries/AdaptiveFilter.pm

```diff
@@ -0,0 +1,210 @@
+    my $trust_w;
+    my $ad;
+    my $ads = [];    # used on build stage only
+    my $mad;
+    my $mads;
+    my $wsum = 0;
+    my $csum = 0;
+    my $vol;
+    my $_accepted;
+
+    # flag initicates, that we still need to accumulate enough time series
+    # before doing actual filtering.
+    my $build = 1;
+
+    # resulting adaptive filter function
+    my $fn = sub {
+        my ($epoch, $spot) = @_;
+
+        # operating on nature log
+        $spot = log $spot;
+
+        # prevent loopback window overgrow
+        while (@$lookback > $lookback_capacity or @$lookback > $floor and $lookback->[0][0] < $epoch - $lookback_period) {
+            my $leaving = shift @$lookback;
+            --$lookback_rejected unless $leaving->[3];
+        }
+        my $accepted;
+
+        if ($build) {
+            # always accept the incoming value
+            ($trust_w, $accepted) = (1, 1);
+
+            # gather absolute differences
+            unless (@$lookback < $build_up_count) {
+                $ad = abs($spot - $lookback->[-$build_up_count][1]) / sqrt($build_up_count);
+                push @$ads, $ad;
+            }
+
+            # build condition: the current tick is 60s newer
+            # and we have enough absolute differences
+            if (    @$minute
+                and $minute->[0] < $epoch - 60
+                and @$ads >= $build_up_count)
+            {
+                $build = 0;
+                my @ads = sort { $a <=> $b } @$ads;
+                my $cut = int(@ads / $build_up_count);
+                @ads  = @ads[$cut .. ($build_up_count - 1) * $cut];
+                $mad  = sum(@ads) / @ads;
+                $mads = [($mad) x scalar(@$decay_speeds)];
+            }
+        } else {
+            # prevent $minute array overgrow
+            while (@$minute and $minute->[0] < $epoch - 60) {
+                shift @$minute;
+            }
+            my $density = 60 / (1 + @$minute);    # the number "1", beacuse we account the current value too
+            my $ha      = $wsum / $csum;
+            my $vol     = $mad / $sqrt2pi;
+            my $diff    = abs($spot - $ha);
+
+            # ther there is zero-difference, we accept current piece of data;
+            # otherwise, if it is zero-volatility (flat), we reject it.
+            my $reject =
+                $diff
+                ? ($vol ? ($diff / $vol) : $reject_criterium + 1)
+                : 0;
+
+            $accepted = !($reject > $reject_criterium);
+            $trust_w = 1 / (1 + ($reject / $reject_criterium)**8);
+
+            if (not $accepted and $lookback_rejected > $cap * @$lookback) {
+                ($accepted, $trust_w) = (1, 1);
+            }
+            $ad = abs($spot - $_accepted->[-$build_up_count][1]) / sqrt($build_up_count);
+            if ($ad) {
+                for my $idx (0 .. @$decay_speeds - 1) {
+                    $mads->[$idx] = $_mad->($mads->[$idx], $ad, $trust_w, $density, $decay_speeds->[$idx]);
```

Comment:
Here is the inlined version of the function. Not much more complicated but a little more efficient.

```perl
my $mu = $trust_w * (1 - exp(-$density * $decay_speeds->[$idx]));
$mads->[$idx] = (1 - $mu) * $mads->[$idx] + $mu * $ad;
```